primitives/statement-store/src/lib.rs

```diff
@@ -0,0 +1,484 @@
+			Some(Proof::Sr25519 { signature, signer }) => {
+				let to_sign = self.signature_material();
+				let signature = sp_core::sr25519::Signature(*signature);
+				let public = sp_core::sr25519::Public(*signer);
+				if signature.verify(to_sign.as_slice(), &public) {
+					SignatureVerificationResult::Valid(signer.clone())
+				} else {
+					SignatureVerificationResult::Invalid
+				}
+			},
+			Some(Proof::Ed25519 { signature, signer }) => {
+				let to_sign = self.signature_material();
+				let signature = sp_core::ed25519::Signature(*signature);
+				let public = sp_core::ed25519::Public(*signer);
+				if signature.verify(to_sign.as_slice(), &public) {
+					SignatureVerificationResult::Valid(signer.clone())
+				} else {
+					SignatureVerificationResult::Invalid
+				}
+			},
+			Some(Proof::Secp256k1Ecdsa { signature, signer }) => {
+				let to_sign = self.signature_material();
+				let signature = sp_core::ecdsa::Signature(*signature);
+				let public = sp_core::ecdsa::Public(*signer);
+				if signature.verify(to_sign.as_slice(), &public) {
+					let sender_hash =
+						<sp_runtime::traits::BlakeTwo256 as sp_core::Hasher>::hash(signer);
+					SignatureVerificationResult::Valid(sender_hash.into())
+				} else {
+					SignatureVerificationResult::Invalid
+				}
+			},
+		}
+	}
+
+	/// Calculate statement hash.
+	#[cfg(feature = "std")]
+	pub fn hash(&self) -> [u8; 32] {
+		self.using_encoded(hash_encoded)
+	}
+
+	/// Returns a topic by topic index.
+	pub fn topic(&self, index: usize) -> Option<Topic> {
+		if index < self.num_topics as usize {
+			Some(self.topics[index].clone())
+		} else {
+			None
+		}
+	}
+
+	/// Returns decryption key if any.
+	pub fn decryption_key(&self) -> Option<DecryptionKey> {
+		self.decryption_key.clone()
+	}
+
+	/// Convert to internal data.
+	pub fn into_data(self) -> Option<Vec<u8>> {
+		self.data
+	}
+
+	/// Get a reference to the statement proof, if any.
+	pub fn proof(&self) -> Option<&Proof> {
+		self.proof.as_ref()
+	}
+
+	/// Return encoded fields that can be signed to construct or verify a proof
+	fn signature_material(&self) -> Vec<u8> {
+		self.encoded(false)
+	}
+
+	/// Return a copy of this statement with proof removed
+	pub fn strip_proof(&mut self) {
+		self.proof = None;
+	}
+
+	/// Set statement proof. Any existing proof is overwritten.
+	pub fn set_proof(&mut self, proof: Proof) {
+		self.proof = Some(proof)
+	}
+
+	/// Set topic by index.
+	pub fn set_topic(&mut self, index: usize, topic: Topic) {
+		if index <= 4 {
+			self.topics[index] = topic;
+			self.num_topics = self.num_topics.max(index as u8 + 1);
+		}
+	}
+
+	/// Set decryption key.
+	pub fn set_decryption_key(&mut self, key: DecryptionKey) {
+		self.decryption_key = Some(key);
+	}
+
+	/// Set unencrypted statement data.
+	pub fn set_plain_data(&mut self, data: Vec<u8>) {
+		self.data = Some(data)
+	}
+
+	fn encoded(&self, with_proof: bool) -> Vec<u8> {
```

Comment:
Here we use custom encoding that encodes as `Vec<Field>` rather than a struct with fixed fields.  The statement object is supposed to be used by external API users. It also travels over the network and stored in the database. It should be possible to add new fields without breaking compatibility for all existing users/UIs.  In fact, the current representation in the rust code as a struct with fixed fields is just an optimization for in-memory storage. It could have been defined as `struct Statement(Vec<Field>)` and the derived encoding would be correct. 
As for pros/cons in general, encoding a struct would be a bit shorter as it would omit vector length prefix and enum IDs for each field.